usd_qtpy/render_util/playblast.py

```diff
@@ -0,0 +1,166 @@
+        raise ValueError("Invalid/Empty filepath for rendering")
+
+    # ensure right complexity object is picked.
+    # the internal _RefinementComplexity.value is used to set rendering quality
+    if isinstance(complexity,str):
+        # ensure key correctness
+        complexity = complexity.lower() # set all to lowercase
+        complexity = complexity.title() # Uppercase Each Word (In Case Of "Very High")
+        preset_names = getComplexityLevels()
+        if complexity not in preset_names:
+            raise ValueError(f"Value: {complexity} entered for complexity is not valid")
+        
+        complex_level = Complex.fromName(complexity)
+    elif isinstance(complexity,int):
+        complexity = min(max(complexity,0),3) # clamp to range of 0-3, 4 elements
+        complex_level = Complex._ordered[complexity]
+
+    complex_level = complex_level.value
+
+    # validate render engine
+    if not checkRenderEngineName(renderer):
+        raise ValueError(f"Render engine arguement invalid")
+    renderer = getRenderPlugin(renderer)
+
+    # TEMP: pick first found camera
+    if not camera:
+        camera = next(findCameras(stage), None)
+        if not camera:
+            # Same procedure as default for pxr.UsdAppUtils.cameraArgs.py
+            path = Sdf.Path(UsdUtils.GetPrimaryCameraName())
+            camera = UsdAppUtils.GetCameraAtPath(stage, path)
+
+    if colormode not in getColorArgs():
+        raise ValueError("Color correction mode specifier is invalid.")
+
+    # Set up OpenGL FBO to write to within Widget
+    # Actual size doesn't matter
+    # it does need to be stored in a variable though, otherwise it'll be collected
+    ogl_widget = _setupOGLWidget(width,width) 
+
+    # Create FrameRecorder
+    frameRecorder = UsdAppUtils.FrameRecorder()
+    frameRecorder.SetRendererPlugin(renderer)
+    frameRecorder.SetImageWidth(width) # Only width is needed, heigh will be computer from camera properties.
+    frameRecorder.SetComplexity(complex_level)
+    frameRecorder.SetColorCorrectionMode(colormode)
+    #frameRecorder.SetIncludedPurposes(["default","render","proxy","guide"]) # set to all purposes for now.
+
+    # Use Usds own frame specification parser
+    # The following are examples of valid FrameSpecs:
+    # 123 - 101:105 - 105:101 - 101:109x2 - 101:110x2 - 101:104x0.5
+    frame_iterator = FrameSpecIterator(frames)
+
+    if not frame_iterator:
+        frame_iterator = [Usd.TimeCode.EarliestTime()]
+
+    for timeCode in frame_iterator:
+        currentframe = outputpath.format(frame = timeCode.GetValue())
+        print(currentframe)
+        try:
+            print("attempting render")
+            frameRecorder.Record(stage, camera, timeCode, currentframe)
+            print("Done!")
+        except Tf.ErrorException as e:
+            logging.error("Recording aborted due to the following failure at time code {0}: {1}".format(timeCode, str(e)))
```

Comment:
```suggestion
            logging.error("Recording aborted due to the following failure at time code %s: %s", timeCode, e)
```

If you format with a logger use `%s` and add in the variables to be formatted as extra arguments. See the suggestion.

This makes it so that the string is only formatted IF the error is handled by a logger handler - if not, then it doesn't pay for the time spent formatting the variables in the string. Errors are more likely handled by default, but still it's good optimization practice.

If you _need_ to format ahead of time since you want to format something complex nicely you should do so by checking if the logger is enabled for that logging level:
```python
        if log.isEnabledFor(logging.DEBUG):
            log.debug("Registered font families: %s", ", ".join(loaded_fonts))
```
Usually however only done for much heavier formatting.

Reply:
noted, I pretty much lifted this from usdrecord.
Will be fixed in next push